test/CMakeLists.txt

```diff
@@ -3,6 +3,8 @@
 find_package(Threads REQUIRED)
 include(CheckCXXCompilerFlag)
 
+include_directories("../src")
```

Comment:
this should not be necessary. We don't want the tests to have blanket access to the private parts of src.

for the same reason, you should be using visibility = ["//visibility:private"] on most of the cc_library targets above.

Reply:
A fair number of the tests #include "../src/<whatever>" so the encapsulation is actually pretty bad. Since this was tests as part of the package I wasn't particularly concerned with the model here, but it's a reasonable concern.

As far as the visibility:private, isn't that the default in the build file? When I was trying to make use of this from an external dependency, I had to add the explicit visibility:public to the one rule I wanted to make visible.

Reply:
ah, you're right. i wasn't sure how they'd set the default.

they include _some things_ from ../src, but don't have full access. i don't know enough about bazel, but would it be better to add a BUILD in src, a BUILD in test, and manage the visibility of certain src/ libraries to test/ that way?